proposals/0000-net-storage-reads-improvements.md

```diff
@@ -84,14 +85,17 @@ If `skipValue` is `true` for a `Key`, then the value associated with this key is
 
 If `includeDescendants` is `true` for a `Key`, then the replier must also include in the proof all keys that are descendant of the given key (in other words, its children, children of children, children of children of children, etc.). It must do so even if `key` itself doesn't have any storage value associated to it. The values of all of these descendants are replaced with their hashes if `skipValue` is `true`, similarly to `key` itself.
 
+The optional `onlyKeysAfter` field can provide a lower bound for the keys contained in the proof. The responder must not include in its proof any node whose key is inferior or equal to the value in `onlyKeysAfter`. If no `onlyKeysAfter` is provided, the response must start with the root node of the trie.
```

Comment:
That seems like an implementation detail, and IMO the response must specifically **not** contain `[1, 2, 15]` the second time.
If the requester uses `onlyKeysAfter` as intended, then there should be no overlap whatsoever between responses.

However you're pointing out a real problem, which is that `onlyKeysAfter` can't represent nibbles.

Reply:
agree it is implementation detail (did approve in the current form).

> must specifically not contain [1, 2, 15]

it is just not possible unless you have fix length keys: previous iteration stopped at node at 1, 2, 15 but you cannot switch to a sibling before you assert there is no node bellow this key.
In fact you can read the node see there is no children, then not include it in the proof, but it means that the emitter of the query also keep in memory that there is no children to the node, so basically this info should be part of the restart key, it sounds quite complex and would save generally one node (can be a bit more if your key looks like 15, 15, 15,...).

> However you're pointing out a real problem, which is that onlyKeysAfter can't represent nibbles

I feel like it would be confusing to put nibbles in the query, we only use keys everywhere and user of the rpc should not need to know anything about the state storage.

Reply:
> but it means that the emitter of the query also keep in memory that there is no children to the node

Yes, that's intended

> I feel like it would be confusing to put nibbles in the query, we only use keys everywhere and user of the rpc should not need to know anything about the state storage.

That's not the RPC layer, that's the networking protocol.

The concept of nibbles is deeply tied with the trie and with Merkle proofs. You can't do anything with the proof you receive in the response without knowing what nibbles are.

Reply:
make sense, I still think it is too complicated.
if I want to do what you suggest, I would need:
```rust
struct Cursor {
  key: Vec<u8>,
  padded: bool, // indicate if last nibble of key is a padding
  next_child: Option<u8>, // indicate iteration next item or parent if none
}
```

Reply:
ah I get where you're going, `next_child` don t need to be an option as if it is parent, we already got the node in previous query.
And if you define next child as the last nibble of key you don t need the field. Then you also not repeat the parent nodes in the proof.
Would work, is stateless from the server point of view, not from the client point of view. Not really compatible with compact proof (or it also cannot be stateless from the server point of view).
Still sounds a bit complex to me.

Reply:
Actually compact local to a reply would work.
Would be ok then (just need to define the OnlyKeysAfter as nibbles (add a boolean to indicate if padded)).

Actually when reading "The responder must not include in its proof any node whose key is inferior or equal to the value in `onlyKeysAfter`." I was thinking of value key not node key.
So this means all parent key will not be include, that s a bit fun as the proof will jump in sequence, but implementable.
From the client side it means keeping a stack of node in memory and passing the cursor, finally I understand what you did propose.

We can just add the padding boolean to `OnlyKeysAfter` and we re good.